rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,214 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
+	robotPose *spatialmath.Pose
+}
+
+const (
+	sigmaLevel        = 7    // level of precision for stdev calculation (determined through experimentation)
+	imageHeight       = 1080 // image height
+	imageWidth        = 1080 // image width
+	missThreshold     = 0.49 // probability limit below which the associated point is assumed to be free space
+	hitThreshold      = 0.55 // probability limit above which the associated point is assumed to indicate an obstacle
+	pointRadius       = 1    // radius of pointcloud point
+	robotMarkerRadius = 5    // radius of robot marker point
+)
+
+// PointCloudToRGBD creates an image of a pointcloud in the XZ plane, scaling the points to a standard image
+// size. It will also add a red marker to the map to represent the location of the robot. The returned depthMap
+// is unused and so will always be nil.
+func (ppRM *ParallelProjectionOntoXZWithRobotMarker) PointCloudToRGBD(cloud pointcloud.PointCloud,
+) (*rimage.Image, *rimage.DepthMap, error) {
+	meta := cloud.MetaData()
+
+	if cloud.Size() == 0 {
+		return nil, nil, errors.New("projection point cloud is empty")
+	}
+
+	meanStdevX, meanStdevZ, err := calculatePointCloudMeanAndStdevXZ(cloud)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	maxX := math.Min(meanStdevX.mean+float64(sigmaLevel)*meanStdevZ.stdev, meta.MaxX)
+	minX := math.Max(meanStdevX.mean-float64(sigmaLevel)*meanStdevZ.stdev, meta.MinX)
+	maxZ := math.Min(meanStdevZ.mean+float64(sigmaLevel)*meanStdevZ.stdev, meta.MaxZ)
+	minZ := math.Max(meanStdevZ.mean-float64(sigmaLevel)*meanStdevZ.stdev, meta.MinZ)
+
+	// Change the max and min values to ensure the robot marker can be represented in the output image
+	var robotMarker spatialmath.Pose
+	if ppRM.robotPose != nil {
+		robotMarker = *ppRM.robotPose
+		maxX = math.Max(maxX, robotMarker.Point().X)
+		minX = math.Min(minX, robotMarker.Point().X)
+		maxZ = math.Max(maxZ, robotMarker.Point().Z)
+		minZ = math.Min(minZ, robotMarker.Point().Z)
+	}
+
+	// Calculate the scale factors
+	scaleFactor := calculateScaleFactor(maxX-minX, maxZ-minZ)
+
+	// Add points in the pointcloud to a new image
+	var pointColor rimage.Color
+	im := rimage.NewImage(imageWidth, imageHeight)
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		x := int(math.Round((pt.X - minX) * scaleFactor))
+		y := int(math.Round((pt.Z - minZ) * scaleFactor))
+
+		// Adds a point to an image using the value to define the color. If no value is available,
+		// the default color of white is used.
+		if x >= 0 && x < imageWidth && y >= 0 && y < imageHeight {
+			pointColor, err = getColorFromProbabilityValue(data)
+			if err != nil {
+				return false
+			}
+			im.Circle(image.Point{X: x, Y: y}, pointRadius, pointColor)
+		}
+		return true
+	})
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Add a red robot marker to the image
+	if ppRM.robotPose != nil {
+		x := int(math.Round((robotMarker.Point().X - minX) * scaleFactor))
```

Comment:
If the robot marker is outside of the standard deviation does this still do the right thing?